File: algorithms/15_ruby/sorting/merge_sort.rb
def merge_sort(arr)
  return arr if arr.length <= 1
  m = arr.length / 2
  merge(merge_sort(arr[0...m]), merge_sort(arr[m..]))
end
def merge(l, r)
  res = []; i = j = 0
  while i < l.length && j < r.length
    res << (l[i] <= r[j] ? l[i += 1; i-1] : r[j += 1; j-1])
  end
  res + l[i..] + r[j..]
end

File: algorithms/16_kotlin/sorting/MergeSort.kt
fun <T : Comparable<T>> mergeSort(arr: List<T>): List<T> {
    if (arr.size <= 1) return arr
    val m = arr.size / 2
    return merge(mergeSort(arr.subList(0,m)), mergeSort(arr.subList(m,arr.size)))
}
fun <T : Comparable<T>> merge(l: List<T>, r: List<T>): List<T> {
    val res = mutableListOf<T>(); var i=0; var j=0
    while (i<l.size&&j<r.size) if(l[i]<=r[j]) res+=l[i++] else res+=r[j++]
    return res+l.subList(i,l.size)+r.subList(j,r.size)
}